nixos/modules/installer/netboot/netboot.nix

```
@@ -72,32 +76,33 @@ with lib;
     # Closures to be copied to the Nix store, namely the init
     # script and the top-level system configuration directory.
     netboot.storeContents =
-      [ config.system.build.toplevel ];
+      [ config.system.build.bootStage2 ];
 
     # Create the squashfs image that contains the Nix store.
     system.build.squashfsStore = pkgs.callPackage ../../../lib/make-squashfs.nix {
       storeContents = config.netboot.storeContents;
     };
 
-
-    # Create the initrd
-    system.build.netbootRamdisk = pkgs.makeInitrdNG {
-      inherit (config.boot.initrd) compressor;
-      prepend = [ "${config.system.build.initialRamdisk}/initrd" ];
-
-      contents =
-        [ { object = config.system.build.squashfsStore;
-            symlink = "/nix-store.squashfs";
-          }
-        ];
-    };
+    # Append the store to the initrd
+    boot.initrd.prepend =
+    let
+      nixStoreCpio = pkgs.makeInitrdNG {
+        inherit (config.boot.initrd) compressor;
+        contents =
+          [ { object = config.system.build.squashfsStore;
+              symlink = "/nix-store.squashfs";
+            }
+          ];
+        };
+    in
+      [ "${nixStoreCpio}/initrd" ];
 
     system.build.netbootIpxeScript = pkgs.writeTextDir "netboot.ipxe" ''
       #!ipxe
       # Use the cmdline variable to allow the user to specify custom kernel params
       # when chainloading this script from other iPXE scripts like netboot.xyz
-      kernel ${pkgs.stdenv.hostPlatform.linux-kernel.target} init=${config.system.build.toplevel}/init initrd=initrd ${toString config.boot.kernelParams} ''${cmdline}
-      initrd initrd
+      kernel ${pkgs.stdenv.hostPlatform.linux-kernel.target} init=${config.system.build.toplevel}/init initrd=${config.system.build.initialRamdisk}/${config.system.boot.loader.initrdFile} ${toString config.boot.kernelParams} ''${cmdline}
```

Is this right? The kernel wouldn't know about the store path for the initrd, right?

Did you mean this?

```suggestion
      kernel ${pkgs.stdenv.hostPlatform.linux-kernel.target} init=${config.system.build.toplevel}/init initrd=${config.system.boot.loader.initrdFile} ${toString config.boot.kernelParams} ''${cmdline}
```

---

You're absolutely right